File: backend/puzzles/engine/rules/sudoku.py
from typing import Optional

from puzzles.engine.games import PuzzleEngineBase
from puzzles.engine.rules.base import ValidationResult, RuleDefinition


def no_duplicate_numbers_in_rows() -> RuleDefinition:
    """
    Generate a rule to validate that no number appears more than once in any row.

    :return: A rule function that validates row uniqueness constraints
    """

    def rule(engine: PuzzleEngineBase) -> Optional[ValidationResult]:
        state = engine.board_state
        rows, cols = engine.rows, engine.cols

        violating_cells = []

        for r in range(rows):
            seen = {}
            for c in range(cols):
                value = state[r * cols + c]

                # Skip empty cells (assuming 0 or similar represents empty)
                if value == 0 or value is None:
                    continue

                if value in seen:
                    # Mark both the original and duplicate cells as violations
                    violating_cells.append({"row": r, "col": seen[value]})
                    violating_cells.append({"row": r, "col": c})
                else:
                    seen[value] = c

        if not violating_cells:
            return None

        return ValidationResult(locations=violating_cells, rule_type="row_duplicate_violation")

    return RuleDefinition(rule)


def no_duplicate_numbers_in_cols() -> RuleDefinition:
    """
    Generate a rule to validate that no number appears more than once in any column.

    :return: A rule function that validates column uniqueness constraints
    """

    def rule(engine: PuzzleEngineBase) -> Optional[ValidationResult]:
        state = engine.board_state
        rows, cols = engine.rows, engine.cols

        violating_cells = []

        for c in range(cols):
            seen = {}
            for r in range(rows):
                value = state[r * cols + c]

                # Skip empty cells (assuming 0 or similar represents empty)
                if value == 0 or value is None:
                    continue

                if value in seen:
                    # Mark both the original and duplicate cells as violations
                    violating_cells.append({"row": seen[value], "col": c})
                    violating_cells.append({"row": r, "col": c})
                else:
                    seen[value] = r

        if not violating_cells:
            return None

        return ValidationResult(locations=violating_cells, rule_type="col_duplicate_violation")

    return RuleDefinition(rule)


def no_duplicate_numbers_in_boxes() -> RuleDefinition:
    """
    Generate a rule to validate that no number appears more than once in any box.
    Automatically determines box size as square root of board side length.
    Works with 4x4 (2x2 boxes) and 9x9 (3x3 boxes) Sudoku boards.

    :return: A rule function that validates box uniqueness constraints
    """

    def rule(engine: PuzzleEngineBase) -> Optional[ValidationResult]:
        state = engine.board_state
        rows, cols = engine.rows, engine.cols

        # Calculate box size as square root of board side length
        import math

        box_size = int(math.sqrt(rows))

        violating_cells = []

        # Iterate through each box
        for box_row in range(0, rows, box_size):
            for box_col in range(0, cols, box_size):
                seen = {}

                # Check all cells in this box
                for r in range(box_row, min(box_row + box_size, rows)):
                    for c in range(box_col, min(box_col + box_size, cols)):
                        value = state[r * cols + c]

                        # Skip empty cells (assuming 0 or similar represents empty)
                        if value == 0 or value is None:
                            continue

                        if value in seen:
                            # Mark both the original and duplicate cells as violations
                            violating_cells.append(seen[value])
                            violating_cells.append({"row": r, "col": c})
                        else:
                            seen[value] = {"row": r, "col": c}

        if not violating_cells:
            return None

        return ValidationResult(locations=violating_cells, rule_type="box_duplicate_violation")

    return RuleDefinition(rule)

File: frontend/src/services/game/engines/rules/sudoku.ts
import type {
  PuzzleEngine,
  RuleDefinition,
  RuleViolation,
  RuleViolationCell,
} from "@/services/game/engines/PuzzleEngine.ts";
import { SudokuCell } from "@/services/game/engines/translator.ts";

export function no_duplicate_numbers_in_rows(): RuleDefinition {
  function rule(engine: PuzzleEngine): RuleViolation | null {
    const state = engine.board_state; // number[][]
    const rows = engine.definition.rows;
    const cols = engine.definition.cols;

    const violating: RuleViolationCell[] = [];

    for (let r = 0; r < rows; r++) {
      const seen = new Map<unknown, number>(); // value → first-col
      for (let c = 0; c < cols; c++) {
        const value = state[r][c];
        if (value === SudokuCell.EMPTY) continue;

        if (seen.has(value)) {
          violating.push({ row: r, col: seen.get(value)! });
          violating.push({ row: r, col: c });
        } else {
          seen.set(value, c);
        }
      }
    }

    return violating.length ? { locations: violating, rule_name: "row_duplicate_violation" } : null;
  }

  return { rule_func: rule };
}

/* ---------- 2. duplicate numbers in columns ---------- */

export function no_duplicate_numbers_in_cols(): RuleDefinition {
  function rule(engine: PuzzleEngine): RuleViolation | null {
    const state = engine.board_state;
    const rows = engine.definition.rows;
    const cols = engine.definition.cols;

    const violating: RuleViolationCell[] = [];

    for (let c = 0; c < cols; c++) {
      const seen = new Map<unknown, number>(); // value → first-row
      for (let r = 0; r < rows; r++) {
        const value = state[r][c];
        if (value === SudokuCell.EMPTY) continue;

        if (seen.has(value)) {
          violating.push({ row: seen.get(value)!, col: c });
          violating.push({ row: r, col: c });
        } else {
          seen.set(value, r);
        }
      }
    }

    return violating.length ? { locations: violating, rule_name: "col_duplicate_violation" } : null;
  }

  return { rule_func: rule };
}

/* ---------- 3. duplicate numbers in boxes ---------- */

export function no_duplicate_numbers_in_boxes(): RuleDefinition {
  function rule(engine: PuzzleEngine): RuleViolation | null {
    const state = engine.board_state;
    const rows = engine.definition.rows;
    const cols = engine.definition.cols;

    const boxSize = Math.sqrt(rows); // 4×4 → 2, 9×9 → 3 …
    if (!Number.isInteger(boxSize)) return null; // non-square board → skip

    const violating: RuleViolationCell[] = [];

    for (let br = 0; br < rows; br += boxSize) {
      for (let bc = 0; bc < cols; bc += boxSize) {
        const seen = new Map<unknown, RuleViolationCell>(); // value → first cell

        for (let r = br; r < br + boxSize; r++) {
          for (let c = bc; c < bc + boxSize; c++) {
            const value = state[r][c];
            if (value === SudokuCell.EMPTY) continue;

            if (seen.has(value)) {
              violating.push(seen.get(value)!);
              violating.push({ row: r, col: c });
            } else {
              seen.set(value, { row: r, col: c });
            }
          }
        }
      }
    }

    return violating.length ? { locations: violating, rule_name: "box_duplicate_violation" } : null;
  }

  return { rule_func: rule };
}
